src/analyses/primitive_clustering.py

```diff
@@ -0,0 +1,200 @@
+                break
+            if path1 not in prim_indexes:
+                prim_indexes[path1] = ind
+                ind += 1
+        for path2 in prim_id_to_paths[prim2]:
+            break
+        prim_indexes[path2] = ind
+        ind += 1
+
+        metric_vars: Dict[str, Dict[str, Any]] = {m.name: {} for m in MetricProblemType}
+        for metric in metric_vars.keys():
+            metric_vars[metric]["distance_matrix"] = np.ones((ind, ind))
+            for i in range(ind):
+                metric_vars[metric]["distance_matrix"][i][i] = 0
+
+            metric_vars[metric]["found_any"] = False
+
+        keys = list(ppcm.keys())
+        while len(keys) > 0:
+            prim_pair = keys.pop(0)
+            run_diff_list = ppcm.pop(prim_pair)
+
+            if len(run_diff_list) == 0:
+                continue
+
+            for metric in metric_vars.keys():
+                metric_vars[metric]["found_iteration"] = False
+                metric_vars[metric]["total"] = 0
+                metric_vars[metric]["num_diffs"] = 0
+
+            while len(run_diff_list) > 0:
+                run_diff = run_diff_list.pop(0)
+
+                metric = run_diff.output_difference_metric.name
+                metric_vars[metric]["found_any"] = True
+                metric_vars[metric]["found_iteration"] = True
+                metric_vars[metric]["total"] += run_diff.output_difference
+                metric_vars[metric]["num_diffs"] += 1
+
+            for metric in metric_vars.keys():
+                if metric_vars[metric]["found_iteration"]:
+                    metric_vars[metric]["avg"] = (
+                        metric_vars[metric]["total"] / metric_vars[metric]["num_diffs"]
+                    )
+
+                    for path1 in prim_id_to_paths[prim_pair[0]]:
+                        break
+                    for path2 in prim_id_to_paths[prim_pair[1]]:
+                        break
+                    metric_vars[metric]["distance_matrix"][prim_indexes[path1]][
+                        prim_indexes[path2]
+                    ] = metric_vars[metric]["avg"]
+
+        for metric in metric_vars.keys():
+            if metric_vars[metric]["found_any"]:
+
+                # The distance matrix provided to squareform() must be symmetric
+                metric_vars[metric]["distance_matrix"] = np.minimum(
+                    metric_vars[metric]["distance_matrix"],
+                    metric_vars[metric]["distance_matrix"].T,
+                )
+
+                labels = list(prim_indexes.keys())
+                # Manually drop the rows/columns of any primitives that aren't close to any others
+                metric_vars[metric]["distance_matrix"], labels = filter_distance_matrix(
+                    metric_vars[metric]["distance_matrix"], labels, self.CLIP_DISTANCE
+                )
+
+                Z = linkage(
+                    squareform(metric_vars[metric]["distance_matrix"]), "average"
+                )
```

Comment:
Purely to satisfy my own curiosity, why was `squareform` needing to be called? It looks like its converting the distance matrix into a 1d condensed distance matrix. It looks like the `linkage` function can take a 2d array of observation vectors.

Also, were you able to verify that the dendrogram is assigning the correct python paths to the correct dendrogram lines?